### PR TITLE
imageproc mllama refactor

### PR DESCRIPTION
**Pull request by pdevine**
This change breaks out the image processing routines into a generic module called `models/imageproc` and also creates a new `models/mllama` model which is specific the the mllama vision processing. There are a few other minor changes such as:
 * Preprocess() now takes an io.Reader instead of sending the byte slice
 * Preprocess() now returns a map[string]any which contains any options to pass back which are specific to the model
 * The mean/standard dev. constants are broken out into package variables

I haven't added an interface for the model, but that should go along with the forward pass and can come in a different PR. We also need to determine what the actual directory structure should look like.

### CODE REVIEW

File: models/mllama/imageproc.go [hunk 1]
```diff
@@ -181,60 +141,61 @@ func PadImage(img image.Image, outputSize, aspectRatio image.Point) image.Image
...
 	}
 
 	return pixelVals
 }
 
-func Preprocess(imageData []byte) ([]float32, int, error) {
-	// todo: need guard in here for bad image data
-
-	// mllama values
+func Preprocess(imageData io.Reader) ([]float32, map[string]any, error) {
```
Comment (pdevine) [inline models/mllama/imageproc.go:181]: This should probably take a slice of images in a future interface.

File: model/mllama/imageproc.go [hunk 2]
```diff
@@ -181,60 +141,61 @@ func PadImage(img image.Image, outputSize, aspectRatio image.Point) image.Image
...
 	}
 
 	return pixelVals
 }
 
-func Preprocess(imageData []byte) ([]float32, int, error) {
-	// todo: need guard in here for bad image data
-
-	// mllama values
+func Preprocess(imageData io.Reader) ([]float32, map[string]any, error) {
```
Comment (jmorganca) [inline model/mllama/imageproc.go:181]: Can be in a follow up but what is `map[string]any` returning? Could we remove this later or at perhaps make it a type?

File: model/pixtral/imageproc.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	"image"
+	_ "image/jpeg"
+	_ "image/png"
+	"io"
+	"math"
+
+	"github.com/ollama/ollama/model/imageproc"
+)
+
+func getNumImageTokens(imageSize, patchSize image.Point) image.Point {
```
Comment (jmorganca) [inline model/pixtral/imageproc.go:14]: prefixing `get` may be extraneous here

```suggestion
func numImageTokens(imageSize, patchSize image.Point) image.Point {
```

File: model/pixtral/imageproc.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	return image.Point{
+		(imageSize.X-1)/patchSize.X + 1,
+		(imageSize.Y-1)/patchSize.Y + 1,
+	}
+}
+
+func getResizeOutputImageSize(img image.Image, longestEdge int, patchSize image.Point) image.Point {
```
Comment (jmorganca) [inline model/pixtral/imageproc.go:21]:
```suggestion
func resizedOutputImageSize(img image.Image, longestEdge int, patchSize image.Point) image.Point {
```

or even `resized`?

**Review by jmorganca — APPROVED** [review]

File: model/mllama/imageproc.go (hunk 2 above)
Comment (pdevine) [reply model/mllama/imageproc.go:181]: it's returning a map of extra return parameters that can be different between models.

> Preprocess() now returns a map[string]any which contains any options to pass back which are specific to the model

in the case of mllama it returns the aspect ratio ID which is unique to the mllama model (and not used by pixtral, qwen2vl, or even llava).

**Pull request merged**